[TDD - 자동차 경주] 2단계 미션 제출합니다.

bigpel66 opened the pull request:
안녕하세요. 1단계 피드백 받은 부분들 위주로 고려하여 리팩토링 진행하였습니다.

1. 뷰와 도메인의 분리에 맞추어, 뷰가 도메인을 의존할 수는 있어도 도메인이 뷰를 의존하지 않는 형태로 작성하려고 노력했습니다.
2. Random 값을 뽑아내는 과정을 protected 상속이 아닌 인터페이스와 구현체를 이용하여 진행하였으며, winner를 뽑아내는 것도 피드백 받은대로 별도 역할을 뽑아내려고 했습니다. winner의 경우 별도 역할을 가진 객체를 만드는 과정에서 전략으로 만들면 좋겠다 싶어서 Random과 마찬가지로 인터페이스 구성 후 구현체 작성하였습니다.
3. 방어적 복사에 대해서 알아보던 중, 단순히 unmodifiableList가 어떤 문제가 있는지 알 수 있었습니다. 주입된 List<Integer>는 외부에서 오는데, 외부에서의 수정이 발생하는 경우에는 Numbers의 콜렉션에도 변동이 생길 수 있다는 것을 알게 되었습니다. 복사 생서과 임시 리스트를 대상으로 unmodifiableList를 이용하여 내외부적으로 불변성을 추구하도록 작성했습니다. 덕분에 객체 내외부 관점에 대해 더 확실히 알 수 있었습니다.

감사합니다.

Commit:
fix: Numbers의 list에 방어적 복사를 적용
- as-is) 단순히 주입 받은 list를 Collections.unmodifiableList로 받았고, 주입 받은 list의 수정이 일어나면 Numbers에도 변경이 발생
- to-be) 값 복사 생성 및 새로운 임시 list를 unmodifiableList로 구성하여, 외부로 주입받은 list에 종속되지 않도록 변경

Commit:
test: Numbers의 list에 방어적 복사가 적용됨을 확인
- 외부로부터 수정되는 작업을 검증
- Numbers 자체가 갖고 있는 list의 수정 방지를 검증

Commit:
refactor: 객체 단위 상수와 객체 상태의 개행 구분

Commit:
refactor: Parser의 parse 메서드의 종속성 제거
- parse 메서드의 반환이 Numbers 타입이 아닌 List<Integer>인 일급 콜렉션을 반환하도록 변경

Commit:
refactor: View로직을 도메인과 분리
- Cars의 자동차들 상태를 일괄적으로 반환하는 메서드를 삭제
- ConsoleView가 Cars를 주입 받아 상태를 그려내는 식으로 변경 (Domain의 의존성을 없애고, View에서는 Domain의 의존성을 설정)
- RacingGame이 레이싱 후, 후처리 로직을 수행할 수 있도록 Consumer타입의 callback을 구성

Commit:
refactor: MVC 구조에 맞추어 패키지 구조 변경
- RacingGame을 Controller로써 정의함에 따라 Controller 메서드의 프로토타입을 변경하고, 턴마다 수행할 행동과 승자 선정 방식을 주입받도록 변경
- Car, Cars를 Model로써 정의함에 따라 model 패키지로 위치 변경
- ConsoleView를 view 패키지로 위치를 변경하고, RacingGame Controller에서 반환된 값을 노출하도록 프로토타입을 변경
- Controller와 View에 해당하는 테스트를 제하고, Model과 Strategy 한정 테스트를 작성

Commit:
refactor: RandomStrategy 작성
- RacingGame의 의존성 주입을 race 호출 시점이 아니라 객체 생성 시로 변경
- Car의 protected 메서드를 private으로 변경하고, 해당 로직을 객체 생성 시점에 Strategy를 주입 받아 수행하도록 변경
- RandomStrategy 인터페이스 정의와 기본 구현체를 정의
- 위 변경사항에 따라 기존 Application 로직과 테스트 코드를 변경
- DefaultRandomStrategy의 iteration 테스트 작성

Commit:
refactor: CarDto 정의
- Car 객체가 뷰 관련 로직을 갖지 않도록 Dto를 별도 구성
- Numbers 및 Cars 대상 list 메서드를 values로 변경
- Numbers 및 Cars 대상 list 멤버 변수를 values로 변경

liquidjoo commented on src/main/java/calculator/Calculator.java:
```diff
@@ -11,9 +11,9 @@ private Calculator() {
         try {
             validate(input);
-            Numbers numbers = Parser.of(input).parse();
+            Numbers numbers = Numbers.of(Parser.of(input).parse());
```
정적팩토리 메서드 패턴은 언제 사용하는 것이 좋을까요?
> 어떤 이유로 사용하게 되었는지 알아보면 도움이 많이 됩니다~

bigpel66 replied:
사용자가 특정 객체를 생성하려는 경우, 인자만 주고서 원하는 타입을 얻을 수 있느냐와 직접적으로 타입을 명시하여 생성하는 점이 다른 거 같습니다. 정.팩.메.를 이용하는 경우엔 후자처럼 구체 타입을 알아야하거나 클래스 API를 뜯어보지 않아도 된다는 점에서 표현력이 높다고 생각했습니다.

혹시 제가 고려하지 못한 점이 있는지 말씀해주시면 감사드리겠습니다.

liquidjoo commented on src/main/java/calculator/Numbers.java:
```diff
@@ -1,34 +1,34 @@
-    public List<Integer> list() {
-        return list;
+    public List<Integer> values() {
+        return values;
```
반환할 때 방어적 복사를 진행해보면 어떤 장점이 있을까요?

bigpel66 replied:
처음엔 Numbers 자체의 불변성만 고려하여 불변 (외부 : 리스트, 내부 : 요소) 리스트를 반환하게 해두었는데요. 그럼에도 다시 한 번 방어적 복사를 말씀해주신 것에 대해 고민해본 결과, 감싸고 있는 콜렉션을 새롭게 구성하면 이용자 입장에서 반환 타입을 유동적으로 사용할 수 있는 장점이 있을 거 같습니다. 기존 객체의 불변성에 영향이 가지 않는 선이라면 이런 유연성을 주는 것이 더 나은 거 같습니다.

liquidjoo commented on src/main/java/calculator/Parser.java:
```diff
@@ -63,12 +64,11 @@ private Matcher getCustomMatcher() {
     }
 
-    public Numbers parse() {
+    public List<Integer> parse() {
```
의존성 제거를 위한 노력 👍

liquidjoo commented on src/main/java/racinggame/controller/RacingGame.java:
```diff
@@ -0,0 +1,41 @@
+        this.winnerStrategy = winnerStrategy;
+    }
+
+    public List<String> race() {
+        for (int i = 0; i < trial; i++) {
+            cars.moveAll();
+            turnCallback.accept(cars);
+        }
+        return winnerStrategy.perform(cars);
+    }
```
책임이 여럿 존재하는 것 같은데요~ 분리해보면 어떨까요?

bigpel66 replied:
for loop부분과 return 부분을 별도로 메서드로 분리하고, race 메서드의 이름을 start로 수정하였습니다. 그리고 분리된 두 메서드를 start 메서드에서 이용하는 식으로 수정하였습니다. start 메서드의 프로토타입은 기존의 race 메서드와 동일하게 유지하였는데, 게임을 마치면 승자가 존재한다는 개념으로 접근하여 반환 타입을 살리고자 했습니다.

liquidjoo commented on src/main/java/racinggame/dto/CarDto.java:
```diff
@@ -0,0 +1,30 @@
+    private final String name;
+    private final int offset;
+
+    public static List<CarDto> from(List<Car> cars) {
```
CarDto 에서 List 를 만들어 반환하는건 조금 어색해보이는데요 :)
별도의 역할을 만들어 보는 것은 어떨까요?

bigpel66 replied:
지적해주신 부분 고민한 결과 Dto가 리스트를 만드는 역할이 맞지 않기도 하고, 이러한 기능이 없어도 되겠다는 생각이 되어 삭제하였습니다.

liquidjoo commented on src/main/java/racinggame/dto/CarDto.java:
```diff
@@ -0,0 +1,30 @@
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    public static CarDto from(Car car) {
+        return new CarDto(car.name(), car.offset());
+    }
+
+    private CarDto(String name, int offset) {
+        this.name = name;
+        this.offset = offset;
+    }
```
정적팩토리 메서드 패턴과 생성자의 차이는 무엇일까요?

bigpel66 replied:
사용자가 특정 객체를 생성하려는 경우, 인자만 주고서 원하는 타입을 얻을 수 있느냐와 직접적으로 타입을 명시하여 생성하는 점이 다른 거 같습니다. 정.팩.메.를 이용하는 경우엔 후자처럼 구체 타입을 알아야하거나 클래스 API를 뜯어보지 않아도 된다는 점에서 표현력이 높다고 생각했습니다.

혹시 제가 고려하지 못한 점이 있는지 말씀해주시면 감사드리겠습니다.

liquidjoo commented on src/main/java/racinggame/model/Car.java:
```diff
@@ -1,26 +1,47 @@
 
-    public static Car of(String owner) {
-        return new Car(owner);
+    private Car(String name, RandomStrategy randomStrategy) {
```
접근제어자 활용 👍

liquidjoo commented on src/main/java/racinggame/model/Car.java:
```diff
@@ -1,26 +1,47 @@
-    protected Car(String name) {
+    private Car(String name, int offset, RandomStrategy randomStrategy) {
         this.name = name;
-        validateOwner();
+        this.offset = offset;
+        this.randomStrategy = randomStrategy;
+        validateRequiredArguments();
```
변수에 데이터를 초기화 하기 전에 validateRequiredArguments 를 먼저 호출해 보면 어떨까요?

bigpel66 replied:
RacingGame에서는 먼저 검사하게 해두고는, 나머지는 유효성 검사를 왜 뒤로 미뤄뒀는지 모르겠네요. 놓친 부분 짚어주셔서 감사합니다. 말씀해주신대로 유효성 검사를 먼저하는 쪽으로 모두 변경하였습니다.

liquidjoo commented on src/main/java/racinggame/model/Cars.java:
```diff
@@ -0,0 +1,34 @@
+
+    private Cars(List<Car> cars) {
+        values = Collections.unmodifiableList(cars);
+        validateCars();
```
유효성 검사를 먼저 해보아요 :)

bigpel66 replied:
유효성을 먼저 검사하여 먼저 자원이 소모되지 않는 형태로 수정하였습니다.

liquidjoo commented on src/main/java/racinggame/model/Cars.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.Set;
+
+public final class Cars {
+    private final List<Car> values;
```
values 라는 네이밍도 좋지만 상태값을 명확히 나타낼 수 있는 이름이 좋지 않을까요~?

bigpel66 replied:
registeredCars라는 이름을 사용하였습니다. 이런 식의 이름을 의도하신 것이 아니라면 어떤 이름들이 사용될 수 있는지 제시해주시면 감사드리겠습니다.

liquidjoo commented on src/test/java/racinggame/model/CarTest.java:
```diff
@@ -13,61 +13,40 @@ public class CarTest {
-                return 4;
-            }
-        };
+        Car car = Car.of("jseo", () -> 4);
```
람다식 표현 👍

liquidjoo reviewed (changes requested):
안녕하세요 종환님 :)
이전 단계의 피드백, 2단계 미션 진행 잘해주셨어요 👍 
몇 가지 코멘트 남겼습니다~ 
확인하고 다시 요청 주세요 🙇

bigpel66 commented:
리뷰해주신 내용 모두 살펴보았습니다.
꼼꼼히 봐주셔서 감사합니다.
내일 중으로 말씀해주신 부분 반영하여 리뷰 요청 남기겠습니다.
늦은 밤에 다시 감사드리고, 잘 주무세요.

Commit:
refactor: 2차 피드백 내용을 반영
- Numbers의 values 메서드의 방어적 복사를 통해 반환의 유연성을 보장
- Parser, Car, Cars의 유효성 검증을 생성자 호출 최상단 시점으로 변경
- RacingGame의 race 메서드의 책임을 분리 (race -> start, start에서는 race 및 winner 메서드를 호출)
- CarDto에서 List를 만드는 책임을 삭제
- Cars의 values 메서드를 상태를 반영한 이름인 registeredCars로 변경
- 위 코드들의 수정에 따라 테스트 코드의 수정

Commit:
fix: final 클래스 명시 누락 수정

liquidjoo commented on src/main/java/calculator/Numbers.java:
```diff
@@ -16,7 +17,7 @@ private Numbers(List<Integer> numbers) {
 
     public List<Integer> values() {
-        return values;
+        return new ArrayList<>(values);
```
피드백 반영 👍

liquidjoo commented on src/main/java/calculator/Parser.java:
```diff
@@ -37,19 +37,19 @@ private Parser(String input) {
-    private void validateInput() {
-        checkInputBlank();
-        checkInputContainsRegex();
+    private void validateInput(String input) {
+        checkInputBlank(input);
+        checkInputContainsRegex(input);
```
피드백 반영 좋습니다 :) 감사합니다~

liquidjoo reviewed (approved):
안녕하세요 종환님 :)
2단계 피드백 반영 잘해주셨어요 💯 
저와 함께 했던 리뷰가 도움이 되었길 빌며
남은 미션들도 화이팅 입니다
수고하셨습니다 🙇
